### PR TITLE
Fix linking for FSSpec

### LINKED ISSUE #134: [Feature]: S3 streaming support via fsspec
### What would you like to see added to HDMF-ZARR?

Support streaming using fsspec 

### Is your feature request related to a problem?

https://github.com/NeurodataWithoutBorders/helpdesk/discussions/56#discussioncomment-7299446 

### What solution would you like?

Add [FSStore](https://zarr.readthedocs.io/en/stable/_modules/zarr/storage.html#FSStore) to support fsspect-based streaming from S3 and/or allow passing of a `zarr.Group` with the parent group as a read-only store for `ZarrIO.__init__(path=...)`

https://github.com/hdmf-dev/hdmf-zarr/blob/70bf35b60ed2c2eaae7b12080bc1f4cc3d89ba3e/src/hdmf_zarr/backend.py#L66C1-L69C47

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/hdmf-dev/hdmf-zarr/blob/dev/.github/CODE_OF_CONDUCT.md)
- [X] Have you checked the [Contributing](https://github.com/hdmf-dev/hdmf-zarr/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this change was not already [requested](https://github.com/hdmf-dev/hdmf-zarr/issues)?

### PR DESCRIPTION
**Pull request by alejoe91**
## Motivation

The `ZarrIO.resolve_ref()` function assumes paths are local. This PR adds a method `is_remote()` to check if the `zarr.store` is `FSStore` and handles remote paths and links correctly.

Fixes #134

## How to test the behavior?
```
from hdmf_zarr import NWBZarrIO

remote_zarr_location = "s3://aind-open-data/ecephys_625749_2022-08-03_15-15-06_nwb_2023-05-16_16-34-55/ecephys_625749_2022-08-03_15-15-06_nwb/ecephys_625749_2022-08-03_15-15-06_experiment1_recording1.nwb.zarr/"

with NWBZarrIO(remote_zarr_location, "r") as io:
    nwbfile = io.read()

print(nwbfile)
```

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf-zarr/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `ruff` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf-zarr/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Comment by alejoe91** [comment]
Should I add a test?

**Comment by alejoe91** [comment]
> Should I add a test?

Added one using the Allen Institute dataset. Not sure if and where you guys want to add it to the CI, since it's currently quite slow. Maybe it would be better to keep this as a place holder and run the test once we have a small test file stored somewhere?

**Comment by oruebel** [comment]
Thanks for taking a stab at this! We can work on getting a small file for testing on DANDI hopefully while we are at SfN. The PR looks overall good to me.

**Comment by oruebel** [comment]
@alejoe91 When running the test locally, I get the following error:

```FAILED tests/unit/test_fsspec_streaming.py::TestFSSpecStreaming::test_fsspec_streaming - botocore.exceptions.NoCredentialsError: Unable to locate credentials```

Is there a way to avoid the need for credentials in the test and/or would using a file on DANDI fix this?

I made a few minor changes to fix flake8 warnings (to avoid unused import warnings) and adding fsspec and s3fs as optional requirements. 

With regard to the time it takes for the test, I think this is Ok for now. I.e., I would be in favor of merging the PR and creating a separate issue to for creating a smaller test file to speed up the test.  I.e., aside from the credential issues with the test, this PR looks good to me.

**Comment by oruebel** [comment]
The credential error also occured in the GitHub action here: 

https://github.com/hdmf-dev/hdmf-zarr/actions/runs/6838973516/job/18596637194?pr=138#step:7:119

**Comment by alejoe91** [comment]
I see! I think I haven't propagated the `storage_options` parameter to all `zarr.open()`! I'll push a fix tomorrow

File: src/hdmf_zarr/backend.py [hunk 1]
```diff
@@ -624,19 +639,21 @@ def resolve_ref(self, zarr_ref):
         else:
             source_file = str(zarr_ref['source'])
         # Resolve the path relative to the current file
-        source_file = os.path.abspath(os.path.join(self.source, source_file))
+        if not self.is_remote():
+            source_file = os.path.abspath(os.path.join(self.source, source_file))
+        else:
+            # get rid of extra "/" and "./" in the path root and source_file
+            root_path = str(self.path)[:-1] if str(self.path).endswith("/") else str(self.path)
+            source_path = str(source_file)[1:] if str(source_file).startswith(".") else str(source_file)
```
Comment (oruebel) [inline src/hdmf_zarr/backend.py]:
```suggestion
            root_path = str(self.path).rstrip("/") 
            source_path = str(source_file).lstrip(".")
```

File: src/hdmf_zarr/backend.py [hunk 2]
```diff
@@ -624,19 +639,21 @@ def resolve_ref(self, zarr_ref):
...
             source_file = str(zarr_ref['source'])
         # Resolve the path relative to the current file
-        source_file = os.path.abspath(os.path.join(self.source, source_file))
+        if not self.is_remote():
+            source_file = os.path.abspath(os.path.join(self.source, source_file))
+        else:
+            # get rid of extra "/" and "./" in the path root and source_file
+            root_path = str(self.path).rstrip("/") 
+            source_path = str(source_file).lstrip(".")
+            source_file = root_path + source_path
```
Comment (oruebel) [inline src/hdmf_zarr/backend.py:648]:
```suggestion
            source_file = os.path.join(root_path, source_path)
```

File: src/hdmf_zarr/backend.py [hunk 3]
```diff
@@ -624,19 +639,21 @@ def resolve_ref(self, zarr_ref):
         else:
             source_file = str(zarr_ref['source'])
         # Resolve the path relative to the current file
-        source_file = os.path.abspath(os.path.join(self.source, source_file))
+        if not self.is_remote():
+            source_file = os.path.abspath(os.path.join(self.source, source_file))
+        else:
+            # get rid of extra "/" and "./" in the path root and source_file
+            root_path = str(self.path).rstrip("/")
```
Comment (oruebel) [inline src/hdmf_zarr/backend.py]:
```suggestion
            root_path = str(self.path).rstrip("/")
```

File: src/hdmf_zarr/backend.py [hunk 4]
```diff
@@ -624,19 +639,21 @@ def resolve_ref(self, zarr_ref):
...
+            source_path = str(source_file).lstrip(".")
+            source_file = os.path.join(root_path, source_path)
```
Comment (oruebel) [inline src/hdmf_zarr/backend.py]:
```suggestion
            source_file = root_path + source_path
```

**Comment by oruebel** [comment]
Just for reference, to help create a test with DANDI:

Example code how to get a S3 URL for Zarr read from DANDI

```
from dandi.dandiapi import DandiAPIClient
from dandi.consts import known_instances

staging_instance = known_instances["dandi-staging"]
dandiset_id = "204919"
version_id = "draft"

client = DandiAPIClient(dandi_instance=staging_instance)
dandiset = client.get_dandiset(dandiset_id=dandiset_id, version_id=version_id)
zarr_asset = dandiset.get_asset_by_path(path="test_read_nwbfile/test2.nwb.zarr")
metadata = zarr_asset.get_metadata()
content = metadata.contentUrl[1]
s3_url = f"s3://dandiarchive{content.path}"
```

**Review by oruebel — APPROVED** [review]

**Review by oruebel — DISMISSED** [review]
Thanks @alejoe91 !

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,10 @@
 # HDMF-ZARR Changelog
 
+## 0.4.1 (?)
+
+### Enhancements
+* Fix linking for FSSpec #138. @alejoe91 [#120](https://github.com/hdmf-dev/hdmf-zarr/pull/138)
```
Comment (oruebel) [inline CHANGELOG.md]:
```suggestion
* Fix linking for FSSpec and support passing of `storage_options` required reading data from S3 #138. @alejoe91 [#120](https://github.com/hdmf-dev/hdmf-zarr/pull/138)
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,10 @@
 # HDMF-ZARR Changelog
 
+## 0.4.1 (?)
```
Comment (oruebel) [inline CHANGELOG.md:3]:
```suggestion
## 0.5.0 (?)
```

Comment (oruebel) [inline CHANGELOG.md]:
```suggestion
## 0.5.0 (Upcoming)
```

**Review by oruebel — APPROVED** [review]

**Pull request merged**